### PR TITLE
Support for Json (de)serialization

### PR DESCRIPTION
**Pull request by joanne-ter-maat**

### CODE REVIEW

**Comment by joanne-ter-maat** [comment]
@ricardoboss any feedback/suggestions for improvement would be welcome.

File: .vscode/settings.json
Comment (ricardoboss) [inline .vscode/settings.json:1]: nit: not a fan of IDE specific settings being checked in...

File: packages/kiota_abstractions/lib/src/case_insensitive_map.dart [hunk 1]
```diff
@@ -109,6 +109,7 @@ class CaseInsensitiveMap<K extends String, V> implements Map<K, V> {
 
   @override
   V putIfAbsent(K key, V Function() ifAbsent) {
+    _originalKeys[normalizeKey(key)] = key;
```
Comment (ricardoboss) [inline packages/kiota_abstractions/lib/src/case_insensitive_map.dart:112]: Nice catch!

Did you add a test for this?

File: packages/kiota_abstractions/lib/src/extensions/request_information_extensions.dart [hunk 2]
```diff
@@ -13,7 +13,7 @@ extension RequestInformationExtensions on RequestInformation {
 
   /// Vanity method to add headers to the request.
   void addHeaders(HttpHeaders headers) {
-    headers.addAll(headers);
+    this.headers.addAll(headers);
```
Comment (ricardoboss) [inline packages/kiota_abstractions/lib/src/extensions/request_information_extensions.dart:16]: Nice catch again!

Test?

File: packages/kiota_abstractions/lib/src/extensions/request_information_extensions.dart [hunk 3]
```diff
@@ -137,16 +137,16 @@ extension RequestInformationExtensions on RequestInformation {
     content = writer.getSerializedContent();
   }
 
-  void configure<T>(void Function(RequestConfiguration)? configurator) {
+  void configure<T extends AbstractQueryParameters>(void Function(RequestConfiguration<T>)? configurator, T Function() createParameters) {
```
Comment (ricardoboss) [inline packages/kiota_abstractions/lib/src/extensions/request_information_extensions.dart:140]: Correct me if I'm wrong, but the `configurator` may be null, but the second parameter can't? IMHO they should both be nullable or neither should.

File: packages/kiota_abstractions/lib/src/extensions/request_information_extensions.dart [hunk 4]
```diff
@@ -137,16 +137,16 @@ extension RequestInformationExtensions on RequestInformation {
...
     }
 
-    final config = RequestConfiguration();
-
+    final config = RequestConfiguration<T>()
+    ..queryParameters = createParameters();
     configurator(config);
-
-    addQueryParameters(config.queryParameters);
+
```
Comment (ricardoboss) [inline packages/kiota_abstractions/lib/src/extensions/request_information_extensions.dart:148]: nit: formatting

File: packages/kiota_abstractions/lib/src/request_configuration.dart [hunk 5]
```diff
@@ -1,13 +1,12 @@
...
+class RequestConfiguration<T extends AbstractQueryParameters> {
   /// The HTTP headers of the request.
   HttpHeaders headers = HttpHeaders();
 
   /// The request options.
   List<RequestOption> options = [];
 
-  /// The request query parameters.
-  Map<String, dynamic> queryParameters = {};
+  late T queryParameters;
```
Comment (ricardoboss) [inline packages/kiota_abstractions/lib/src/request_configuration.dart:11]: I don't this is a good design. It should only be marked `late` if you can guarantee that this value will be set before being accessed.

I'd opt to make this nullable instead.

File: packages/kiota_abstractions/lib/src/serialization/untyped_array.dart [hunk 6]
```diff
@@ -0,0 +1,15 @@
+part of '../../kiota_abstractions.dart';
+
+/// Represents an untyped node with a collection of other untyped nodes.
+class UntypedArray extends UntypedNode {
+
+  /// Constructs an instance from the supplied [collection]
+  UntypedArray(this.collection);
+
+  Iterable<UntypedNode> collection;
```
Comment (ricardoboss) [inline packages/kiota_abstractions/lib/src/serialization/untyped_array.dart:9]: Any reason why not `final`? Could mark the constructor `const`

Comment (ricardoboss) [reply packages/kiota_abstractions/lib/src/serialization/untyped_array.dart:9]: I see most other constructors could be `const` as well

File: packages/kiota_abstractions/lib/src/serialization/untyped_null.dart [hunk 7]
```diff
@@ -0,0 +1,7 @@
+part of '../../kiota_abstractions.dart';
+
+class UntypedNull extends UntypedNode {
+  UntypedNull();
+
+  final dynamic value = null;
```
Comment (ricardoboss) [inline packages/kiota_abstractions/lib/src/serialization/untyped_null.dart:6]: This seems to be unused

File: packages/kiota_abstractions/lib/src/serialization/untyped_node.dart [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+/// Base class for untyped node.
+class UntypedNode implements Parsable {
+    UntypedNode();
+
+    /// Creates a new [UntypedNode] instance of the appropriate class
+    /// based on discriminator value.
+    factory UntypedNode.createFromDiscriminatorValue(ParseNode parseNode) {
+      return UntypedNode();
+    }
```
Comment (ricardoboss) [inline packages/kiota_abstractions/lib/src/serialization/untyped_node.dart:11]: What is this supposed to do?

File: .gitignore [hunk 9]
```diff
@@ -29,8 +29,11 @@ pubspec.lock
 build/
 
 # ignore generated files
-gen/
+**/gen/**
 *.mocks.dart
 
+# Do not ignore generated version
+!**/gen/version.dart
+
```
Comment (ricardoboss) [inline .gitignore:37]: Why? It's generated code and is not supposed to be checked in. It should be generated when publishing a new version, at least that's the idea

File: packages/kiota_http/lib/gen/version.dart
Comment (ricardoboss) [inline packages/kiota_http/lib/gen/version.dart:1]: (said it above, this should not be checked in)

File: packages/kiota_http/lib/src/http_client_request_adapter.dart [hunk 10]
```diff
@@ -80,7 +80,7 @@ class HttpClientRequestAdapter implements RequestAdapter {
   }
 
   void _setBaseUrl(RequestInformation requestInfo) {
-    requestInfo.pathParameters['baseUrl'] = baseUrl;
+    requestInfo.pathParameters['baseurl'] = baseUrl;
```
Comment (ricardoboss) [inline packages/kiota_http/lib/src/http_client_request_adapter.dart:83]: Why this change?

File: packages/kiota_abstractions/lib/src/api_exception.dart [hunk 11]
```diff
@@ -12,13 +12,13 @@ class ApiException implements Exception {
   });
 
   /// The HTTP status code of the response.
-  final int? statusCode;
+  int? statusCode;
```
Comment (ricardoboss) [inline packages/kiota_abstractions/lib/src/api_exception.dart:15]: You wrote in the commit message that we shouldn't use `copyWith` because subclasses loose their values when copying.

The `copyWith` pattern is a commonly used pattern in Dart and I'd like to keep it, ideally. The idea is that subclasses should override the `copyWith` with their custom properties.

Does this not work for the use-cases here?

File: packages/kiota_http/lib/src/http_client_request_adapter.dart [hunk 12]
```diff
@@ -130,12 +130,10 @@ class HttpClientRequestAdapter implements RequestAdapter {
...
-    if (result case Exception exception) {
-      if (exception case final ApiException apiException) {
-        exception = apiException.copyWith(
-          statusCode: statusCodeInt,
-          responseHeaders: headers,
-        );
+    if (result case final Exception exception) {
+    if (exception case final ApiException apiException) {
+        apiException..statusCode = statusCodeInt
+        ..responseHeaders = headers;
```
Comment (ricardoboss) [inline packages/kiota_http/lib/src/http_client_request_adapter.dart:136]: (said it earlier... not a fan of this change)

File: packages/kiota_serialization_json/lib/src/json_parse_node.dart [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+      }
+    }
+    return result;
+  }
+
+  @override
+  Iterable<T> getCollectionOfPrimitiveValues<T>() {
+    final result = <T>[];
+    if (_node is List) {
+      _node.forEach((value) => result.add(value as T));
```
Comment (ricardoboss) [inline packages/kiota_serialization_json/lib/src/json_parse_node.dart:77]: You just use `cast<T>()` here:

```suggestion
      return _node.cast<T>();
```

File: packages/kiota_serialization_json/lib/src/json_parse_node.dart [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+
+    onBeforeAssignFieldValues?.call(item);
+    _assignFieldValues(item);
+    onAfterAssignFieldValues?.call(item);
+
+    return item;
+  }
+
+  void _assignFieldValues<T extends Parsable>(T item) {
+    if (_node is Map) {
```
Comment (ricardoboss) [inline packages/kiota_serialization_json/lib/src/json_parse_node.dart:133]: If you invert this check, you can do an early return

File: packages/kiota_serialization_json/lib/src/json_parse_node.dart [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+    }
+
+    if (node is Map) {
+      final propertiesMap = <String, UntypedNode>{};
+      for (final entry in node.entries) {
+        final fieldKey = entry.key as String;
+        final fieldValue = entry.value;
+        final childNode = JsonParseNode(fieldValue);
+        // childNode.setOnBeforeAssignFieldValues(this.getOnBeforeAssignFieldValues());
+        // childNode.setOnAfterAssignFieldValues(this.getOnAfterAssignFieldValues());
```
Comment (ricardoboss) [inline packages/kiota_serialization_json/lib/src/json_parse_node.dart:187]: This is not working yet?

File: packages/kiota_serialization_json/lib/src/json_parse_node.dart [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+    final boolValue = bool.tryParse(node.toString());
+    if (boolValue != null) {
+      return UntypedBoolean(value: boolValue);
+    }
+
+    if (node is String) {
+      return UntypedString(node);
+    }
+
+    throw ArgumentError.value(node, 'node');
```
Comment (ricardoboss) [inline packages/kiota_serialization_json/lib/src/json_parse_node.dart:213]: nit: Can we add a descriptive message here as well?

File: packages/kiota_serialization_json/lib/src/json_parse_node.dart [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+  }
+
+  @override
+  DateTime? getDateTimeValue() {
+    return _node == null ? null : DateTime.tryParse(_node.toString());
+  }
+
+  @override
+  double? getDoubleValue() {
+    return _node == null ? null : double.tryParse(_node.toString());
```
Comment (ricardoboss) [inline packages/kiota_serialization_json/lib/src/json_parse_node.dart:95]: I'm not 100% certain but I think `jsonDecode` already decodes numbers to their appropriate types. We should test for this and ensure we don't convert a `double` to `string` just to parse it back to a `double`.

Comment (ricardoboss) [reply packages/kiota_serialization_json/lib/src/json_parse_node.dart:95]: Same goes for `int` and `bool`

File: packages/kiota_serialization_json/lib/src/json_serialization_writer.dart
Comment (ricardoboss) [inline packages/kiota_serialization_json/lib/src/json_serialization_writer.dart:1]: I'm sure this took a lot of effort to create but I think Darts `jsonEncode` does a pretty good job of encoding a `Map<String, Object>` to JSON. We should consider _not_ creating our own serializer here.

File: packages/kiota_serialization_json/pubspec.yaml [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+  kiota_abstractions:
+    path: ../kiota_abstractions
+  uuid: ^4.4.2
+
+dev_dependencies:
+  strict: ^2.0.0
+  test: ^1.25.2
+
+
+
```
Comment (ricardoboss) [inline packages/kiota_serialization_json/pubspec.yaml:22]: nit: formatting

**Review by ricardoboss — COMMENTED** [review]
Sorry it took me so long to finally review this!

Thanks for your work on this, looks very good already. A few nits and some design decision questions popped up.

File: packages/kiota_abstractions/lib/src/api_exception.dart (hunk 11 above)
Comment (joanne-ter-maat) [reply packages/kiota_abstractions/lib/src/api_exception.dart:15]: I wasn't aware of this pattern, but gave it a shot. I'm not sure if I used it correctly. To me it seems weird to ignore all the arguments of the copywith method. This is how you intended for it to be used? 
```
class CustomError extends ApiException {
  String? someProperty;
  int? someNumber;

  CustomError({
    this.someNumber,
    this.someProperty,
  });

  @override
  ApiException copyWith({
    int? statusCode,
    String? message,
    Map<String, List<String>>? responseHeaders,
    Iterable<Object?>? innerExceptions,
  }) {
    return CustomError(
      someNumber: someNumber, 
      someProperty: someProperty);
  }
}
```

Comment (ricardoboss) [reply packages/kiota_abstractions/lib/src/api_exception.dart:15]: Not quite. The `copyWith` method is supposed to take in all the parameters the constructor takes in (but nullable, if not already) and passes all of its properties as fallbacks for the parameters given to `copyWith`:

```dart
class Foo {
    const Foo({required this.val});

    final int val;

    Foo copyWith({
        int? val,
    }) {
        return Foo(val: val ?? this.val);
    }
}
```

With `copyWith` you can change a single property while keeping all other values.

I imagined an override to look like this:

```dart
class Bar extends Foo {
    const Bar({
        required super.val,
        required this.mul,
    });

    final int mul;

    @override
    Bar copyWith({
        int? val,
        int? mul,
    }) {
        return Bar(
            val: val ?? this.val,
            mul: mul ?? this.mul,
        );
    }
}
```

**Comment by ricardoboss** [comment]
@joanne-ter-maat @Kees-Schotanus the repository has moved to a new owner, so please remember to update your remote urls!

File: packages/kiota_abstractions/lib/src/request_configuration.dart [hunk 19]
```diff
@@ -2,11 +2,13 @@ part of '../kiota_abstractions.dart';
...
+
+  RequestConfiguration({required this.queryParameters});
   /// The HTTP headers of the request.
   HttpHeaders headers = HttpHeaders();
 
   /// The request options.
   List<RequestOption> options = [];
 
-  late T queryParameters;
+  T queryParameters;
```
Comment (ricardoboss) [inline packages/kiota_abstractions/lib/src/request_configuration.dart:13]: Ideally, you should try to make all properties `final`, because then the constructor can be made `const`, which will help the Dart runtime optimize usages of this object.

File: packages/kiota_abstractions/lib/src/serialization/untyped_array.dart (hunk 6 above)
Comment (Kees-Schotanus) [reply packages/kiota_abstractions/lib/src/serialization/untyped_array.dart:9]: Made all untyped constructors const

File: packages/kiota_abstractions/lib/src/serialization/untyped_node.dart (hunk 8 above)
Comment (Kees-Schotanus) [reply packages/kiota_abstractions/lib/src/serialization/untyped_node.dart:11]: It creates an UntypedNode which is actually used when running json_parse_node_test.
The parameter is not used in this case.
It has the same implementation in dot-net and Java

File: packages/kiota_abstractions/lib/src/serialization/untyped_null.dart (hunk 7 above)
Comment (Kees-Schotanus) [reply packages/kiota_abstractions/lib/src/serialization/untyped_null.dart:6]: I changed this code to not having a value at all
and overriding the getValue() method to always return null.

File: packages/kiota_http/lib/src/http_client_request_adapter.dart (hunk 10 above)
Comment (Kees-Schotanus) [reply packages/kiota_http/lib/src/http_client_request_adapter.dart:83]: You used 'baseurl' everywhere else in the code, so I changed it.
For example in request_information_test.dart:  requestInfo.pathParameters['baseurl'] = 'http://localhost';

File: packages/kiota_abstractions/lib/src/extensions/request_information_extensions.dart (hunk 3 above)
Comment (Kees-Schotanus) [reply packages/kiota_abstractions/lib/src/extensions/request_information_extensions.dart]: Both params can now be null.

File: packages/kiota_abstractions/lib/src/extensions/request_information_extensions.dart (hunk 4 above)
Comment (Kees-Schotanus) [reply packages/kiota_abstractions/lib/src/extensions/request_information_extensions.dart]: Adjusted formatting

File: packages/kiota_abstractions/lib/src/request_configuration.dart (hunk 5 above)
Comment (Kees-Schotanus) [reply packages/kiota_abstractions/lib/src/request_configuration.dart]: Joanne fixed this. The late keyword is gone.

File: .gitignore (hunk 9 above)
Comment (Kees-Schotanus) [reply .gitignore]: My bad. While executing the code we got an error that version.dart was missing and this solved the problem at the time.

File: packages/kiota_http/lib/gen/version.dart
Comment (Kees-Schotanus) [reply packages/kiota_http/lib/gen/version.dart:1]: Joanne has fixed this.

File: packages/kiota_serialization_json/lib/src/json_parse_node.dart (hunk 14 above)
Comment (Kees-Schotanus) [reply packages/kiota_serialization_json/lib/src/json_parse_node.dart]: Done by Joanne

File: packages/kiota_serialization_json/lib/src/json_parse_node.dart (hunk 15 above)
Comment (Kees-Schotanus) [reply packages/kiota_serialization_json/lib/src/json_parse_node.dart]: This should work.

File: packages/kiota_serialization_json/lib/src/json_parse_node.dart (hunk 16 above)
Comment (Kees-Schotanus) [reply packages/kiota_serialization_json/lib/src/json_parse_node.dart]: Done by Joanne

File: packages/kiota_abstractions/lib/src/request_configuration.dart (hunk 19 above)
Comment (Kees-Schotanus) [reply packages/kiota_abstractions/lib/src/request_configuration.dart]: Updated to const constructor by Joanne

File: packages/kiota_serialization_json/lib/src/json_serialization_writer.dart [hunk 20]
```diff
@@ -0,0 +1,278 @@
+part of '../kiota_serialization_json.dart';
+
+class JsonSerializationWriter implements SerializationWriter {
+  final List<String> _buffer = [];
```
Comment (ricardoboss) [inline packages/kiota_serialization_json/lib/src/json_serialization_writer.dart:4]: There is a `StringBuffer` class that efficiently lets you build strings. Maybe we should use that instead https://api.dart.dev/stable/3.5.4/dart-core/StringBuffer-class.html

File: packages/kiota_serialization_form/lib/src/form_serialization_writer.dart [hunk 21]
```diff
@@ -168,12 +168,14 @@ class FormSerializationWriter implements SerializationWriter {
 
       if (additionalValuesToMerge != null) {
         for (final additionalValue in additionalValuesToMerge) {
-          onBeforeObjectSerialization?.call(additionalValue);
-          onStartObjectSerialization?.call(additionalValue, this);
+          if(additionalValue != null){
```
[the remark this answers is not in the document]
Comment (Kees-Schotanus) [reply packages/kiota_serialization_form/lib/src/form_serialization_writer.dart:171]: Condition is reverted.

File: packages/kiota_serialization_json/lib/src/json_serialization_writer.dart (hunk 20 above)
Comment (Kees-Schotanus) [reply packages/kiota_serialization_json/lib/src/json_serialization_writer.dart:4]: @ricardoboss 

The StringBuffer class does not allow for deletions and we need to delete the separator character if it is the last one in the buffer (used in multiple places).
We could use Uint8Buffer as I have done for the MultiPart writer, resulting in code like this:
_buffer.addAll(utf8.encode('"${entry.key}":${entry.value}'));

Just let me know what you think.

**Review by ricardoboss — COMMENTED** [review]
Looks really good. I think this is ready for merging soon!

File: packages/kiota_serialization_json/lib/src/json_serialization_writer.dart (hunk 20 above)
Comment (ricardoboss) [reply packages/kiota_serialization_json/lib/src/json_serialization_writer.dart:4]: I left a new review

**Comment by ricardoboss** [comment]
@joanne-ter-maat @Kees-Schotanus can one of you add yourselves to the `.github/CODEOWNERS` file for the json package?

File: packages/kiota_serialization_form/lib/src/form_serialization_writer.dart
[the remark this answers is not in the document]
Comment (joanne-ter-maat) [reply packages/kiota_serialization_form/lib/src/form_serialization_writer.dart:1]: I've made some changes to the json_serialization_writer to stop using a string buffer.

[the remark this answers is not in the document]
Comment (ricardoboss) [reply packages/kiota_serialization_form/lib/src/form_serialization_writer.dart:1]: Thanks! I think it should be more maintainable and since you wrote so many good tests we can be relatively sure everything still works

File: packages/kiota_abstractions/lib/src/serialization/untyped_string.dart [hunk 22]
```diff
@@ -0,0 +1,7 @@
+part of '../../kiota_abstractions.dart';
+
+class UntypedString extends UntypedNode {
+  const UntypedString(this.value);
+
+  final String value;
```
[the remark this answers is not in the document]
Comment (joanne-ter-maat) [reply packages/kiota_abstractions/lib/src/serialization/untyped_string.dart:6]: As far as I can tell from looking at the abstractions for the other languages, the intention seems to be to have that logic in the serialization writer, so I added it there.

[the remark this answers is not in the document]
Comment (ricardoboss) [reply packages/kiota_abstractions/lib/src/serialization/untyped_string.dart:6]: Okay, I see where you're going with this. Still, in the writer there are these lines in `writeObjectValue`:

```dart
      if (value is UntypedNode) {
        writeUntypedValue(key, value as UntypedNode);
      } else {
        value.serialize(this);
      }
```

`writeUntypedValue` in turn has a switch for more specific types. I'm just wondering why you did this when you could have implemented the `serialize` method from the `Parsable` interface in each type instead. Then you wouldn't have to handle all this in the writer.

[the remark this answers is not in the document]
Comment (joanne-ter-maat) [reply packages/kiota_abstractions/lib/src/serialization/untyped_string.dart:6]: The UntypedNode and its subclasses represent just the value. I don't think you can implement the serialize method there, because you then lose the key, since the key is not known within the UntypedNode subclasses.

[the remark this answers is not in the document]
Comment (ricardoboss) [reply packages/kiota_abstractions/lib/src/serialization/untyped_string.dart:6]: Oh I understand now, thanks for explaining! I forgot the `Untyped*` classes live in the abstractions layer. Makes sense that they are independent of the serializer used.

File: packages/kiota_http/lib/src/kiota_client_factory.dart [hunk 23]
```diff
@@ -20,7 +20,7 @@ class KiotaClientFactory {
           return retryCodes.contains(response.statusCode);
         },
       ),
-      userAgent: 'kiota_http/$packageVersion',
+      userAgent: 'kiota_http/0.0.1',
```
[the remark this answers is not in the document]
Comment (ricardoboss) [reply packages/kiota_http/lib/src/kiota_client_factory.dart:23]: For now, resolved by https://github.com/kiota-community/dart_kiota/pull/45

**Review by ricardoboss — APPROVED** [review]
Looking good! Just fix the conflicts and this should be ready. Great work!
